### PR TITLE
Wheel plowing approximation in ODEPhysics

### PR DESCRIPTION
**Pull request by scpeters**
As noted in https://github.com/osrf/gazebo/issues/3085, the distribution of contact forces can be quite different for a wheel driving on rigid terrain and deformable terrain.  Contact for rigid wheel on rigid terrain can be modeled with a single contact point and normal direction perpendicular to the surface. For a rigid wheel on deformable terrain, the contact occurs over a larger area. The contact pressure distribution can be approximated as an equivalent point force at the centroid of the pressure distribution. At rest, a wheel on a deformable surface may sink into the surface with a symmetric distribution of normal contact pressure, such that the equivalent normal force vector is parallel to the undisturbed surface normal, which is vertical in the illustration below.

wheel_plowing

While rolling, the sinkage of the wheel may cause a plowing effect, in which the terrain material is bunched up in the direction of travel and compacted in its wake. This can cause a significant forward bias in the distribution of normal contact pressure and an equivalent normal force vector that is inclined relative to the undisturbed surface normal. In the illustration above, the equivalent normal force opposes the motion of the wheel, requiring tractive torque to be applied in order to maintain a steady velocity. Even when flat, the plowing effect means that driving on a deformable surface is energetically equivalent to driving uphill on a rigid surface.

This pull request implements an approximation of longitudinal wheel plowing effects by displacing contact points and inclining the contact normals to replicate the behavior illustrated in the figure above. This is implemented in `ODEPhysics::Collide`, which is a callback function called for each pair of colliding bodies. To enable the plowing effect, special SDFormat tags must be added to the terrain (`<gz:plowing_terrain />`) and the wheels (`<gz:plowing_wheel />`). The wheel must specify `<fdir1>` in its `<surface>`, and the `<gz:plowing_wheel>` tag must also contain `<max_degrees>` and `<saturation_velocity>` and may optionally contain `<deadband_velocity>`. See the `plowing_effect_*` models in `test/models` for examples.

### CODE REVIEW

**Comment by adityapande-1995** [comment]
Left : Vehicle using the plowing effect, right : without the plowing effect
image

Wheels sinking into the ground : 
image

**Comment by adityapande-1995** [comment]
To run the example : ``gazebo --verbose worlds/plowing_effect_demo.world``

**Comment by scpeters** [comment]
I've updated this and marked it ready for review

**Review by WilliamLewww — APPROVED** [review]
A minor warning:
```
[Err] [InsertModelWidget.cc:403] Missing model.config for model "/gazebo-ws/shared/gazebo/test/models/test_nested_urdf"
[Err] [InsertModelWidget.cc:403] Missing model.config for model "/gazebo-ws/shared/gazebo/test/models/testdb"
```

but looks good to me!

File: test/models/plowing_effect_ground_plane/model.config [hunk 1]
```diff
@@ -0,0 +1,16 @@
+<?xml version="1.0"?>
+
+<model>
+  <name>Ground Plane with plowing effect</name>
+  <version>1.0</version>
+  <sdf version="1.5">model.sdf</sdf>
+
+  <author>
+    <name>Nate Koenig</name>
+    <email>nate@osrfoundation.org</email>
```
Comment (scpeters) [inline test/models/plowing_effect_ground_plane/model.config:10]: @adityapande-1995 I'll add you as a co-author on these new models

File: gazebo/physics/ode/ODEPhysics.cc [hunk 2]
```diff
@@ -1692,3 +1800,41 @@ bool ODEPhysics::GetParam(const std::string &_key, boost::any &_value) const
   }
   return true;
 }
+
+/// \brief A mathematical function that represents a saturation function
+/// that increases linearly until reaching a constant value with the addition
+/// of a deadband near zero. THe shape of this function for positive input
```
Comment (scpeters) [inline gazebo/physics/ode/ODEPhysics.cc]: capitalization: `THe` -> `The`

File: gazebo/physics/ode/ODEPhysics.cc [hunk 3]
```diff
@@ -1692,3 +1800,41 @@ bool ODEPhysics::GetParam(const std::string &_key, boost::any &_value) const
...
+    --+-------------------------- input
+      |<----> dead-band
+
+ \endverbatim */
+double saturation_deadband(
+    double _maxOutput, double _deadband, double _slope, double _input)
+{
+  if (std::abs(_input) <= std::abs(_deadband))
+  {
+    return 0;
```
Comment (scpeters) [inline gazebo/physics/ode/ODEPhysics.cc]: return 0.0

File: gazebo/physics/ode/ODEPhysics.cc [hunk 4]
```diff
@@ -1161,6 +1164,111 @@ void ODEPhysics::Collide(ODECollision *_collision1, ODECollision *_collision2,
...
+      plowingMaxAngle.SetDegree(wheelSdf->Get<double>(kPlowingMaxDegrees));
+      double plowingSaturationVelocity =
+        wheelSdf->Get<double>(kPlowingSaturationVelocity);
+      // Use deadband velocity = 0 if parameter is not set
+      double plowingDeadbandVelocity =
+        wheelSdf->Get<double>(kPlowingDeadbandVelocity, 0.0).first;
+
+      // compute slope
+      double slope = plowingMaxAngle.Radian() /
+          (plowingSaturationVelocity - plowingDeadbandVelocity);
```
Comment (scpeters) [inline gazebo/physics/ode/ODEPhysics.cc]: check that saturation velocity is larger than deadband velocity

File: test/models/plowing_effect_ground_plane/model.config (hunk 1 above)
Comment (scpeters) [reply test/models/plowing_effect_ground_plane/model.config:10]: https://github.com/osrf/gazebo/pull/3164/commits/cd9a5869193cbf59ba5e4b311856aa20668e0a20

File: gazebo/physics/ode/ODEPhysics.cc (hunk 2 above)
Comment (scpeters) [reply gazebo/physics/ode/ODEPhysics.cc]: https://github.com/osrf/gazebo/pull/3164/commits/cd9a5869193cbf59ba5e4b311856aa20668e0a20

File: gazebo/physics/ode/ODEPhysics.cc (hunk 3 above)
Comment (scpeters) [reply gazebo/physics/ode/ODEPhysics.cc]: https://github.com/osrf/gazebo/pull/3164/commits/cd9a5869193cbf59ba5e4b311856aa20668e0a20

File: gazebo/physics/ode/ODEPhysics.cc (hunk 4 above)
Comment (scpeters) [reply gazebo/physics/ode/ODEPhysics.cc]: this ended up being a bigger change, as I moved the SDF parameter parsing to a static helper function in ODECollision and added parameter checks there

https://github.com/osrf/gazebo/pull/3164/commits/dd64a81cc396cc1ef7adf27fd167867bb0bc1570

**Comment by scpeters** [comment]
tests have been kindly provided by @deepanshubansal01 in https://github.com/osrf/gazebo/pull/3229, and I have made some small adjustments:

* https://github.com/osrf/gazebo/pull/3164/commits/b6cf816760253292a4832bd29d3515e2563311e6: minor refactoring of test code and models
* https://github.com/osrf/gazebo/pull/3164/commits/ecedcc198c2bb26444928a7691018d0c632b82bd: adjustments to tricycle on deformable terrain test: update parameters to ensure all plowing states are reached, fix one of the calculations, and improve console messaging
* https://github.com/osrf/gazebo/pull/3164/commits/5432fe58a916aeefa6b2b82116bb863b63f6f320: improve contact parameters in plowing_effect_test.world to improve contact point stability

**Review by adityapande-1995 — APPROVED** [review]
Tests look good ! I feel we should document in the examples / code that the plowing effect only applies in the direction of ``fdir1`` for now.

**Review by deepanshubansal01 — APPROVED** [review]
Looks good to me as well.

**Comment by scpeters** [comment]
> Tests look good ! I feel we should document in the examples / code that the plowing effect only applies in the direction of `fdir1` for now.

To be precise, the plowing effect applies in the direction normal to `fdir1` and the surface normal. If `fdir1` is the wheel's lateral direction, then the plowing effect will apply in the wheel's longitudinal direction.  I'll go ahead and add some notes about this in the example model / world

**Comment by scpeters** [comment]
> > Tests look good ! I feel we should document in the examples / code that the plowing effect only applies in the direction of `fdir1` for now.
> 
> To be precise, the plowing effect applies in the direction normal to `fdir1` and the surface normal. If `fdir1` is the wheel's lateral direction, then the plowing effect will apply in the wheel's longitudinal direction. I'll go ahead and add some notes about this in the example model / world

I've added documentation in several recent commits; let me know what you think @adityapande-1995 and @deepanshubansal01

**Comment by adityapande-1995** [comment]
> > > Tests look good ! I feel we should document in the examples / code that the plowing effect only applies in the direction of `fdir1` for now.
> > 
> > 
> > To be precise, the plowing effect applies in the direction normal to `fdir1` and the surface normal. If `fdir1` is the wheel's lateral direction, then the plowing effect will apply in the wheel's longitudinal direction. I'll go ahead and add some notes about this in the example model / world
> 
> I've added documentation in several recent commits; let me know what you think @adityapande-1995 and @deepanshubansal01

Looks good to me !

**Comment by deepanshubansal01** [comment]
> > > Tests look good ! I feel we should document in the examples / code that the plowing effect only applies in the direction of `fdir1` for now.
> > 
> > 
> > To be precise, the plowing effect applies in the direction normal to `fdir1` and the surface normal. If `fdir1` is the wheel's lateral direction, then the plowing effect will apply in the wheel's longitudinal direction. I'll go ahead and add some notes about this in the example model / world
> 
> I've added documentation in several recent commits; let me know what you think @adityapande-1995 and @deepanshubansal01

Looks good to me as well !

**Pull request merged**